Fix Bug 1648832 - Hostname is configured as local-ipv4 in OpenStack

bysnupy opened the pull request:
Fix: https://bugzilla.redhat.com/show_bug.cgi?id=1648832

The `ip_var` would have `local-ipv4` of `ec_compat`, so it can be `10.0.1.1,10.0.2.2` as `hostname`(`f_var`).
So we should split the string as `IP address` before assigning as `hostname`.

For example.
~~~
            facts['network'][f_var] = metadata['ec2_compat'][ip_var]
~~~

vrutkovs commented:
/cc @tomassedovic

sdodson commented:
/cc @tzumainn

tzumainn commented:
I'm not very familiar with this code; where are the values for the ec2_compat metadata coming from... ? Is it from provisioned openstack resources, or elsewhere?

It does look like a similar workaround was implemented in this file elsewhere already: https://github.com/openshift/openshift-ansible/blob/master/roles/openshift_facts/library/openshift_facts.py#L303

But I'm not sure how to reproduce this issue in the first place.

bysnupy commented:
* "ec2_compat" from here : https://github.com/openshift/openshift-ansible/blob/release-3.9/roles/openshift_facts/library/openshift_facts.py#L1468-L1479
  * And the value is provided from OpenStack meta-data service as "x.x.x.x,y.y.y.y" format. (maybe...)
~~~
def guess_host_provider(self):
    """ Guess the host provider
        Returns:
            dict: The generated default facts for the detected provider
    """
...
        elif re.search(r'OpenStack', product_name):
            provider = 'openstack'
            metadata_url = ('http://169.254.169.254/openstack/latest/'
                            'meta_data.json')
            metadata = get_provider_metadata(metadata_url, True, None,
                                             True)

            if metadata:
                ec2_compat_url = 'http://169.254.169.254/latest/meta-data/'
                metadata['ec2_compat'] = get_provider_metadata(
                    ec2_compat_url
                )
~~~

* The code lines are some confusing, because the variable name and key name in dictionary data collection was same.
~~~
    facts['zone'] = metadata['availability_zone']
    # if the string can be separated with ',', then first array(IP) is assigned to "local-ipv4 VARIABLE", NOT KEY name, 
    # and the original value was not changed in ec2_compat.local-ipv4.
    local_ipv4 = metadata['ec2_compat']['local-ipv4'].split(',')[0]
    # network.ip is assigned to above IP address.
    facts['network']['ip'] = local_ipv4
    facts['network']['public_ip'] = metadata['ec2_compat']['public-ipv4']

    # "f_var" is "hostname", "h_var" is "hostname" and "ip_var" is "local-ipv4" as KEY name in first loop.
    for f_var, h_var, ip_var in [('hostname', 'hostname', 'local-ipv4'),
                                 ('public_hostname', 'public-hostname', 'public-ipv4')]:
        try:
            # This condition become false, because the IP(10.0.1.1) resolved from hostname and ec2_compat.local-ipv4("10.0.1.1,10.0.2.2") is not equal.
            if socket.gethostbyname(metadata['ec2_compat'][h_var]) == metadata['ec2_compat'][ip_var]:
                facts['network'][f_var] = metadata['ec2_compat'][h_var]
            # So, network.hostname is assigned as ec2_compat.local-ipv4("10.0.1.1,10.0.2.2") and it's root cause of this bug.
            else:
                facts['network'][f_var] = metadata['ec2_compat'][ip_var]
        except socket.gaierror:
            facts['network'][f_var] = metadata['ec2_compat'][ip_var]
~~~

* You can test simple code using temporary values as follows: 
~~~
import socket
"""
Usage: python test.py
"""

metadata={'ec2_compat': {}}
facts={'network': {}}
metadata['ec2_compat']['local-ipv4'] = "10.0.1.1,10.0.2.2"
metadata['ec2_compat']['public-ipv4'] = ""
metadata['ec2_compat']['hostname'] = "master1.example.com"
metadata['ec2_compat']['public-hostname'] = "master1.example.com"


local_ipv4 = metadata['ec2_compat']['local-ipv4'].split(',')[0]
facts['network']['ip'] = local_ipv4
facts['network']['public_ip'] = metadata['ec2_compat']['public-ipv4']

for f_var, h_var, ip_var in [('hostname', 'hostname', 'local-ipv4'),
                             ('public_hostname', 'public-hostname', 'public-ipv4')]:
    try:
        if socket.gethostbyname(metadata['ec2_compat'][h_var]) == metadata['ec2_compat'][ip_var]:
            facts['network'][f_var] = metadata['ec2_compat'][h_var]
        else:
            facts['network'][f_var] = metadata['ec2_compat'][ip_var]
    except socket.gaierror:
        facts['network'][f_var] = metadata['ec2_compat'][ip_var]

print facts
~~~

And I added more information at the BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1648832

tzumainn commented:
@bysnupy Thanks for the added information! Quick question - are you running the playbooks in openshift-ansible/playbooks/openstack ? Or is this being installed some other way? I'm still not entirely sure how the metadata service is getting those values.

bysnupy commented:
@tzumainn 

> Quick question - are you running the playbooks in openshift-ansible/playbooks/openstack ?

No, I'm not. This issue occurred during ` /usr/share/ansible/openshift-ansible/playbooks/prerequisites.yml`.

> Or is this being installed some other way? I'm still not entirely sure how the metadata service is getting those values.

No, it is not. It's usual steps of installation. And the value was collected by `/usr/share/ansible/openshift-ansible/playbooks/init/main.yml` via `openshift_facts.py`, it's part of `prerequisites.yml`. 

1. `/usr/share/ansible/openshift-ansible/playbooks/init/main.yml` has collected various `facts` on the target hosts. (I think the related `OpenStack` data is collected at this time.)

2. `openshift.common.hostname` is configured by above tasks, so the ansible logs were as follows.
[ https://github.com/openshift/openshift-ansible/blob/release-3.9/playbooks/init/validate_hostnames.yml#L2-L11  ]
~~~
TASK [Query DNS for IP address of 10.0.1.1,10.0.2.2] ****************************************
ok: [master1.example.com] => {"changed": false, "cmd": "getent ahostsv4 10.0.1.1,10.0.2.2 | head -n 1 | awk '{ print $1 }'", "delta": "0:00:04.123456", "end": "2018-11-07 06:28:45.514534", "failed": false, "failed_when_result": false, "rc": 0, "start": "2018-11-07 06:28:41.279799", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}
~~~
3. Next task was failed with following syntax error messages.
[ https://github.com/openshift/openshift-ansible/blob/release-3.9/playbooks/init/validate_hostnames.yml#L13-L28 ]
~~~
TASK [Validate openshift_hostname when defined] ****************************************************
fatal: [master1.example.com]: FAILED! => {"failed": true, "msg": "The task includes an option with an undefined variable. The error was: 'openshift_hostname' is undefined\n\nThe error appears to have been in '/usr/share/ansible/openshift-ansible/playbooks/init/validate_hostnames.yml': line 13, column 5, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n  - name: Validate openshift_hostname when defined\n    ^ here\n\nexception type: <class 'ansible.errors.AnsibleUndefinedVariable'>\nexception: 'openshift_hostname' is undefined"}
~~~

tzumainn commented:
@sdodson this fix looks perfectly reasonable to me, as it looks like a similar fix had already been added elsewhere in the file - however it's not something I'm able to reproduce. Do you know who may have been working on openstack-related installs prior to 3.10?

sdodson commented:
/ok-to-test

sdodson commented:
@tzumainn I'm not sure where the original openstack work came from. @cooktheryan may know.

cooktheryan commented:
Tagging in @e-minguez @rlopez133 who may be more familiar with this work than i

e-minguez commented:
I'm not familiar with the code involving the PR (according to git blame it has been untouched for ~4 years), but I have some doubts:

- Does the metadata server always answer with the same IP order?
- I'm not sure about OCP working fine with 2 network interfaces. There are some BZs about that[1][2][3] and a KCS[4] on how to properly set the listening interface using node-config

[1] https://bugzilla.redhat.com/show_bug.cgi?id=1346466
[2] https://bugzilla.redhat.com/show_bug.cgi?id=1477989
[3] https://bugzilla.redhat.com/show_bug.cgi?id=1642843
[4] https://access.redhat.com/solutions/3625721

bysnupy commented:
@e-minguez

> Does the metadata server always answer with the same IP order?

Exactly, I also worry about it, refer the BZ[0]. In fact this issue is part of it. (This issues occurred during installation on `OpenStack`.)

And `OpenStack` cloud provider seems to be implemented by the information based on `OpenStack metadata service API`, so I think the `IP` addresses order is controlled by `OpenStack` side. (The IP addresses order depend on `OpenStack`) I have no idea whether or not `IP` addresses order should be handled by ansible playbooks of `OpenShift`. If the `IP` addresses order does not keep, why we implement `OpenStack` cloud provider based on it.

> I'm not sure about OCP working fine with 2 network interfaces. There are some BZs about that[1][2][3] and a KCS[4] on how to properly set the listening interface using node-config

This topic is interesting for me and I want to discuss more, but it's out of focus about this issue. 
@sdodson might answer about it. :^)

The issues is configured hostname as `x.x.x.x,y.y.y.y` by not converting the string to IP address.

[0] [Does the metadata server always answer with the same IP order?](https://bugzilla.redhat.com/show_bug.cgi?id=1650030)

sdodson commented:
@tzumainn do you know if the ip addresses coming back from the metadata apis are deterministicly ordered?

nekop commented:
This seems to be a regression when reverting commits. The code L328-334 takes care of the comma separated IPs:

https://github.com/openshift/openshift-ansible/blob/dfb95684ef5f636f0595e311e0c5169ee5c1ec92/roles/openshift_facts/library/openshift_facts.py#L328-L334

But the following block is updated after that and it doesn't care about the comma separated IPs.

From the openstack side fix, it seems the IPs are coming in a fixed order, <instanceIP, Controller_Local_IP>. And it's <instanceIP> only after the fix.

tzumainn commented:
/lgtm

bysnupy commented:
@sdodson - would you PTAL ?

sdodson commented:
/approve

sdodson commented:
Is this relevant in 3.10 and 3.11?

bysnupy commented:
Thanks a lot for review @sdodson , yes it is. Their versions also would be required this patch.

sdodson commented:
/cherrypick release-3.10
/cherrypick release-3.11

sdodson commented:
/cherrypick release-3.11

nekop commented:
@sdodson cherrypick for 3.11 not working for some reason?

sdodson commented:
/cherrypick release-3.11

sdodson commented:
> @sdodson cherrypick for 3.11 not working for some reason?

yeah, flakey bots, thanks for following up

therevoman commented:
Just ran into an issue with this change.  My OpenStack environment returns a list object which does not have a split function.
